hyperspy/signal.py

```diff
@@ -3288,99 +3288,227 @@ def save(
+
+                Process multiple files and save in different format:
+
+                >>> import hyperspy.api as hs
+                >>> from pathlib import Path
+                >>>
+                >>> input_folder = Path("input_data/")
+                >>> output_folder = Path("processed_data/")
+                >>>
+                >>> for file_path in input_folder.glob("*.hspy"):
+                ...     s = hs.load(file_path)
+                ...     # Process the signal...
+                ...     s = s.remove_background()
+                ...     # Save in new location with different format
+                ...     s.save(output_folder, file_format="msa")
 
         """
+        # Check for conflicting parameters
+        if extension is not None and file_format is not None:
+            raise ValueError(
+                "Cannot specify both 'extension' and 'file_format' parameters. "
+                "Please use only 'file_format' as 'extension' is deprecated."
+            )
+
+        # Deprecation warning for extension parameter
+        if extension is not None:
+            warnings.warn(
+                "The 'extension' parameter is deprecated in HyperSpy 2.4 and will be removed in HyperSpy 3.0. "
+                "Please use 'file_format' instead.",
+                FutureWarning,
+                stacklevel=2,
+            )
+
         if filename is None:
             if self.tmp_parameters.has_item(
                 "filename"
             ) and self.tmp_parameters.has_item("folder"):
+                # Determine the extension to use
+                if self.tmp_parameters.has_item("extension"):
+                    file_ext = self.tmp_parameters.extension
+                elif file_format is not None:
+                    # Get the default extension for the file format from rsciio
+                    try:
+                        from hyperspy.io import _format_name_to_reader
+
+                        writer = _format_name_to_reader(file_format)
+                        file_ext = (
+                            "." + writer["file_extensions"][writer["default_extension"]]
+                        )
+                    except (ValueError, KeyError):
+                        # If format not found in rsciio, use the file_format as extension
+                        file_ext = "." + file_format
+                else:
+                    file_ext = ".hspy"  # Default extension
+
                 filename = Path(
-                    self.tmp_parameters.folder, self.tmp_parameters.filename
-                )
-                extension = (
-                    self.tmp_parameters.extension if not extension else extension
+                    self.tmp_parameters.folder,
+                    self.tmp_parameters.filename + file_ext,
                 )
+                # Don't override extension if it was explicitly provided
+                if extension is None and self.tmp_parameters.has_item("extension"):
```

Comment:
Try to simplify the code by parsing `extension` before - see comment above.

Reply:
See reasoning as above.